### PR TITLE
Twisha/stat util

### PR DESCRIPTION
**Pull request by twisha-shah99**
Aggregator:
image

Trainer(s):
image

### CODE REVIEW

File: lib/python/examples/fwdllm/trainer/forward_training/tc_transformer_trainer_distribute.py [hunk 1]
```diff
@@ -303,6 +302,17 @@ def train_model(self, device=None, logging_state=None):
                     x = batch[1].to(device, non_blocking=True)
                     labels = batch[4].to(device, non_blocking=True)
 
+                    if epoch == 0 and batch_idx == 0 and hasattr(self, "base_trainer"):
```
Comment (gdadlaney) [inline lib/python/examples/fwdllm/trainer/forward_training/tc_transformer_trainer_distribute.py]: As discussed, please verify if we need the epoch & batch_idx checks, given that we've decided to send the stat_utility along with the gradients in every iteration.

Comment (twisha-shah99) [reply lib/python/examples/fwdllm/trainer/forward_training/tc_transformer_trainer_distribute.py]: Do you have a full run log? Can verify through that first

File: lib/python/examples/fwdllm/trainer/forward_training/tc_transformer_trainer_distribute.py [hunk 2]
```diff
@@ -303,6 +302,17 @@ def train_model(self, device=None, logging_state=None):
...
+                        with torch.no_grad():
+                            pred = self.model(x)
+                            if hasattr(pred, "logits"):
+                                logits = pred.logits
+                            elif isinstance(pred, (tuple, list)):
+                                logits = pred[0]
+                            else:
+                                logits = pred
+                            loss = self.base_trainer.oort_loss(logits, labels.view(-1), epoch=0, batch_idx=0, reduction="mean")
```
Comment (gdadlaney) [inline lib/python/examples/fwdllm/trainer/forward_training/tc_transformer_trainer_distribute.py]: Shouldn't it be `reduction="sum"` here? We're anyway dividing by the loss `_batch_size` in normalize_stat_utility(), so do we intend to divide the sum of losses by `_batch_size` twice?

Comment (twisha-shah99) [reply lib/python/examples/fwdllm/trainer/forward_training/tc_transformer_trainer_distribute.py]: It shouldn't matter right? -> if you check the oort_loss implementation we are concerned with per sample loss rather than mean or summation. Current implementation defaults to mean if we dont pass the parameter and it is just using (mean/sum) for calculating loss and not stat_utility.

File: lib/python/examples/fwdllm/trainer/forward_training/tc_transformer_trainer_distribute.py (hunk 1 above)
Comment (gdadlaney) [reply lib/python/examples/fwdllm/trainer/forward_training/tc_transformer_trainer_distribute.py]: Yep, you can look at any of 3 trainer logs files in `/home/dgarg39/gaurav/flame/lib/python/examples/fwdllm/expts/run_tc_expts/log/new/benchmarked_runs` on Shepherd. Because the size of these files are in GBs, it'll be much quicker for you to just check them on the Server. The trainer log file names start with `test_trainer` and you do a Cmd+F for `round: 2` to jump to the second round.

File: lib/python/examples/fwdllm/trainer/forward_training/tc_transformer_trainer_distribute.py (hunk 2 above)
Comment (gdadlaney) [reply lib/python/examples/fwdllm/trainer/forward_training/tc_transformer_trainer_distribute.py]: Ah, I now see that this just influences the log & not the `stat_utility` calculation. Thanks for the clarification. Resolving this comment because the `stat_utility` calculation does use the sum of squares loss correctly

Comment (dhruvsgarg) [inline lib/python/examples/fwdllm/trainer/forward_training/tc_transformer_trainer_distribute.py]: since this is inside the if condition, it will only happen iff epoch is 0 and batch_idx is also 0? It might be happening in the oort trainer as well but I don't remember why we are doing this. I thought loss would be calculated after each data-bin is processed

File: lib/python/examples/fwdllm/trainer/forward_training/tc_transformer_trainer_distribute.py [hunk 3]
```diff
@@ -378,7 +388,11 @@ def train_model(self, device=None, logging_state=None):
                     gc.collect()
                     torch.cuda.empty_cache()
                     self.log_memory(f"epoch{epoch}_batch{batch_idx}_end", device)
-
+                
+                if epoch == 0 and hasattr(self, "base_trainer"):
+                    self.base_trainer.normalize_stat_utility(epoch)
```
Comment (dhruvsgarg) [inline lib/python/examples/fwdllm/trainer/forward_training/tc_transformer_trainer_distribute.py:393]: Havent checked the normalize_stat_util logic here but it makes sense for stat util normalization to happen based off the first loss value generated

File: lib/python/flame/mode/horizontal/syncfl/fwdllm_trainer.py [hunk 4]
```diff
@@ -672,9 +678,9 @@ def normalize_stat_utility(self, epoch) -> None:
         Normalize statistical utility of a trainer based on the size of the
         trainer's datset, at epoch 1.
         """
-        if epoch == 1:
-            self._stat_utility = len(self.train_loader.dataset) * math.sqrt(
-                self._stat_utility / len(self.train_loader.dataset)
+        if epoch == 0:
```
Comment (dhruvsgarg) [inline lib/python/flame/mode/horizontal/syncfl/fwdllm_trainer.py]: why are we changing in two places from epoch 1 to epoch 0? Need to be careful about how/when the stat util normalization is happening

**Review by dhruvsgarg — COMMENTED** [review]
Please check the comments
